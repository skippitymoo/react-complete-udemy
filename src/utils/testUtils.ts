/**
 * React Testing Library uses jsdom, which currently does not support HTMLDialogElement
 * https://github.com/testing-library/react-testing-library/issues/1106#issuecomment-1209419325
 * https://github.com/jsdom/jsdom/issues/3294#issuecomment-1268330372
 */
export const setUpModalContainer = (): HTMLDivElement => {
  document.body.replaceChildren();
  const modalDiv = document.createElement('div');
  modalDiv.setAttribute('id', 'modal');
  const modalContainer = document.body.appendChild(modalDiv);

  const showMock = jest.fn().mockImplementation(function mock(this: HTMLDialogElement) {
    this.setAttribute('open', '');
    this.open = true;
  });

  HTMLDialogElement.prototype.show = showMock;

  HTMLDialogElement.prototype.showModal = showMock;

  HTMLDialogElement.prototype.close = jest.fn().mockImplementation(function mock(
    this: HTMLDialogElement,
  ) {
    this.open = false;
  });

  return modalContainer;
};
