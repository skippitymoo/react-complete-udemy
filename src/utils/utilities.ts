export const formatCurrency = (val: number): string => {
  return new Intl.NumberFormat('en-GB', {
    style: 'currency',
    currency: 'GBP',
  }).format(val);
};
