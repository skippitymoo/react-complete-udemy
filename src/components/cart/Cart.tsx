import { forwardRef, useImperativeHandle, useRef, useState } from 'react';
import { CartItem as CartItemType, CartProps } from '../../shared/types/menu.types';
import { CartItem } from './CartItem';
import { formatCurrency } from '../../utils/utilities';
import { Button } from '../ui/Button';
import { Dialog, DialogRef } from '../ui/Dialog';

type ClickType = 'close' | 'order';

export const Cart = forwardRef<DialogRef, CartProps>(function CartModal(
  { initialCartItems, onCartChange, onOrder },
  ref,
) {
  const [cart, setCart] = useState(initialCartItems);
  const dialogRef = useRef<DialogRef>(null);

  useImperativeHandle(ref, (): DialogRef => dialogRef.current!);

  const amountChangeHandler = (mealId: string, changeBy: number): void => {
    setCart((prev: CartItemType[]): CartItemType[] => {
      const updatedCart = [...prev];
      const item = updatedCart.find((i) => i.meal.id === mealId);
      if (item) {
        item.amount += changeBy;
      }

      const totalItems = updatedCart.reduce((prevVal, currVal) => prevVal + currVal.amount, 0);
      onCartChange(totalItems);

      return updatedCart;
    });
  };

  const actionHandler = (clickType: ClickType): void => {
    if (clickType === 'order') {
      // TODO: order the items in the cart
      onOrder();
    } else if (clickType === 'close') {
      dialogRef.current?.close();
    }
  };

  const totalPrice = cart.reduce((prev, curr) => {
    return prev + curr.amount * curr.meal.price;
  }, 0);

  return (
    <Dialog
      ref={dialogRef!}
      aria-labelledby='Cart Summary'
      aria-describedby='Adjust items in your cart and order when ready'
      onClose={actionHandler.bind(null, 'close')}
    >
      <section className='cart'>
        <ul>
          {cart.map((item: CartItemType) => (
            <li key={item.meal.id} className='cart__list-item'>
              <CartItem
                amount={item.amount}
                meal={item.meal}
                onAmountChange={amountChangeHandler}
              />
            </li>
          ))}
        </ul>
        <div className='cart__total'>
          <span>Total Amount</span>
          <span>{formatCurrency(totalPrice)}</span>
        </div>
        <div className='cart__actions'>
          <Button variant='light' onClick={actionHandler.bind(null, 'close')}>
            Close
          </Button>
          <Button onClick={actionHandler.bind(null, 'order')}>Order</Button>
        </div>
      </section>
    </Dialog>
  );
});
