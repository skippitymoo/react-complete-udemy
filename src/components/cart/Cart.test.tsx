import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import { Cart } from './Cart';
import { formatCurrency } from '../../utils/utilities';

expect.extend(toHaveNoViolations);

describe('Cart', () => {
  it('renders', async () => {
    const { container } = render(<Cart onClose={() => {}} onOrder={() => {}} />);

    expect(await axe(container)).toHaveNoViolations();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the total', async () => {
    const total = (22.99 * 1) + (16.5 * 4); // TODO: this will change when hard coded menu in Cart is changed
    render(<Cart onClose={() => {}} onOrder={() => {}} />);

    const renderedTotal = screen.getByText(formatCurrency(total));

    expect(renderedTotal).toBeInTheDocument();
  });

  it('calls the "onClose" handler supplied', () => {
    const onClose = jest.fn();

    render(<Cart onClose={onClose} onOrder={() => {}} />);

    const renderedAddButton = screen.getByRole('button', { name: 'Close' });
    renderedAddButton.click();

    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('calls the "onOrder" handler supplied', () => {
    const onOrder = jest.fn();

    render(<Cart onClose={() => {}} onOrder={onOrder} />);

    const renderedAddButton = screen.getByRole('button', { name: 'Order' });
    renderedAddButton.click();

    expect(onOrder).toHaveBeenCalledTimes(1);
  });
});
