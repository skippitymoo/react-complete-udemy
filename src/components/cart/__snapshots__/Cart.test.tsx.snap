// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders 1`] = `
<dialog
  aria-describedby="Adjust items in your cart and order when ready"
  aria-labelledby="Cart Summary"
  class="dialog"
  open=""
>
  <section
    class="cart"
  >
    <ul>
      <li
        class="cart__list-item"
      >
        <div
          class="cart-item"
        >
          <div
            class="cart-item__details"
          >
            <h3
              class="cart-item__title"
            >
              Sushi
            </h3>
            <span
              class="cart-item__price"
            >
              £22.99
            </span>
            <span
              class="chip cart-item__amount"
            >
              x 1
            </span>
          </div>
          <div
            class="cart-item__actions"
          >
            <button
              class="btn btn--light btn--small"
            >
              −
            </button>
            <button
              class="btn btn--light btn--small"
            >
              +
            </button>
          </div>
        </div>
      </li>
      <li
        class="cart__list-item"
      >
        <div
          class="cart-item"
        >
          <div
            class="cart-item__details"
          >
            <h3
              class="cart-item__title"
            >
              Schnitzel
            </h3>
            <span
              class="cart-item__price"
            >
              £16.50
            </span>
            <span
              class="chip cart-item__amount"
            >
              x 4
            </span>
          </div>
          <div
            class="cart-item__actions"
          >
            <button
              class="btn btn--light btn--small"
            >
              −
            </button>
            <button
              class="btn btn--light btn--small"
            >
              +
            </button>
          </div>
        </div>
      </li>
    </ul>
    <div
      class="cart__total"
    >
      <span>
        Total Amount
      </span>
      <span>
        £88.99
      </span>
    </div>
    <div
      class="cart__actions"
    >
      <button
        class="btn btn--light"
      >
        Close
      </button>
      <button
        class="btn btn--primary"
      >
        Order
      </button>
    </div>
  </section>
</dialog>
`;
